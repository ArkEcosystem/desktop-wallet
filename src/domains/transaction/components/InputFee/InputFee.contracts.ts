import { Networks } from "@arkecosystem/platform-sdk";
import { Contracts } from "@arkecosystem/platform-sdk-profiles";
import { BigNumber } from "@arkecosystem/platform-sdk-support";

enum InputFeeViewType {
	Simple,
	Advanced,
}

enum InputFeeSimpleValue {
	Slow = "slow",
	Average = "average",
	Fast = "fast",
}

const DEFAULT_SIMPLE_VALUE = InputFeeSimpleValue.Average;
const DEFAULT_VIEW_TYPE = InputFeeViewType.Simple;

type InputFeeSimpleOptions = {
	[key in InputFeeSimpleValue]: {
		label: string;
		displayValue: string;
		displayValueConverted: BigNumber;
	};
};

interface InputFeeAdvancedProperties {
	onChange: (value: string) => void;
	value: string;
	min: string;
	max: string;
	step: number;
	disabled?: boolean;
}

interface InputFeeSimpleProperties {
	options: InputFeeSimpleOptions;
	onChange: (value: InputFeeSimpleValue) => void;
	value: InputFeeSimpleValue;
	ticker: string;
	exchangeTicker: string;
	showConvertedValues: boolean;
	loading?: boolean;
}

interface InputFeeProperties {
	min: string;
	avg: string;
	max: string;
	step: number;
	disabled?: boolean;
	network?: Networks.Network;
	profile?: Contracts.IProfile;
	loading?: boolean;
	viewType?: InputFeeViewType;
	simpleValue?: InputFeeSimpleValue;
	value: string;
	onChange: (value: string) => void;
	onChangeViewType: (value: InputFeeViewType) => void;
	onChangeSimpleValue: (value: InputFeeSimpleValue) => void;
}

export { DEFAULT_SIMPLE_VALUE, DEFAULT_VIEW_TYPE, InputFeeSimpleValue, InputFeeViewType };

export type { InputFeeAdvancedProperties, InputFeeProperties, InputFeeSimpleOptions, InputFeeSimpleProperties };
