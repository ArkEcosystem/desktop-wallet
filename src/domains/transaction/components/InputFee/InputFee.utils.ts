import { BigNumber } from "@arkecosystem/platform-sdk-support";

export enum InputFeeViewType {
	Simple,
	Advanced,
}

export enum InputFeeSimpleValue {
	Slow = "slow",
	Average = "average",
	Fast = "fast",
}

export type InputFeeSimpleOptions = {
	[key in InputFeeSimpleValue]: {
		label: string;
		displayValue: string;
		displayValueConverted: BigNumber;
	};
};
