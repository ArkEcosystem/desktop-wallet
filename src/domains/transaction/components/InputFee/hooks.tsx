import { BigNumber } from "@arkecosystem/platform-sdk-support";
import { useEffect, useRef, useState } from "react";

export type InputFee = {
	display: string;
	value: string;
};

export const useFeeFormat = ({ defaultValue, value, avg }: any) => {
	const toHuman = (inputValue: string | number) => BigNumber.make(inputValue).divide(1e8).toString();
	const defaultFeeValue = value || defaultValue || avg;
	const defaultHuman = toHuman(defaultFeeValue);

	const [fee, setFee] = useState<InputFee>({ display: defaultHuman, value: defaultFeeValue });

	const feeFromValue = (feevalue?: string | number): InputFee => {
		const value = BigNumber.make(feevalue || 0);
		return {
			display: value.divide(1e8).toString(),
			value: value.toString(),
		};
	};

	const updatedFee = useRef(fee);
	useEffect(() => {
		updatedFee.current = feeFromValue(fee?.value);
	}, [fee]);

	useEffect(() => {
		if (!value && !defaultValue) {
			setFee(feeFromValue(value));
		}

		const isFeeOutdated = !BigNumber.make(value).isEqualTo(updatedFee.current.value);
		if (isFeeOutdated) {
			setFee(feeFromValue(value));
		}
	}, [value, defaultValue, updatedFee]);

	return {
		fee,
		toHuman,
		updateFee: setFee,
	};
};
