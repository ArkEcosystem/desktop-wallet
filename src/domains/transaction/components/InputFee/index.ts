export { InputFee } from "./InputFee";
export { InputFeeSimpleValue, InputFeeViewType } from "./InputFee.utils";
