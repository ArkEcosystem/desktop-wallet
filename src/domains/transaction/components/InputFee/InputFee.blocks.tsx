import { Amount } from "app/components/Amount";
import { ButtonGroup, ButtonGroupOption } from "app/components/ButtonGroup";
import { InputRange } from "app/components/Input";
import { Skeleton } from "app/components/Skeleton";
import React from "react";

import { InputFeeAdvancedProperties, InputFeeSimpleProperties, InputFeeSimpleValue } from "./InputFee.contracts";

export const InputFeeAdvanced = ({ onChange, value, min, max, step, disabled }: InputFeeAdvancedProperties) => (
	<InputRange disabled={disabled} value={value} min={+min} max={+max} step={step} onChange={onChange} />
);

export const InputFeeSimple = ({
	options,
	onChange,
	value,
	ticker,
	exchangeTicker,
	showConvertedValues,
	loading,
}: InputFeeSimpleProperties) => (
	<ButtonGroup>
		{Object.entries(options).map(([optionValue, { label, displayValue, displayValueConverted }]) => (
			<ButtonGroupOption
				key={optionValue}
				value={displayValue}
				isSelected={() => optionValue === value}
				setSelectedValue={() => onChange(optionValue as InputFeeSimpleValue)}
			>
				<div className="flex flex-col space-y-2">
					<div>{label}</div>
					{loading ? (
						<Skeleton width={100} className="my-1 h-3" />
					) : (
						<>
							<div className="text-sm">
								{displayValue} {ticker}
							</div>
							{showConvertedValues && (
								<div className="text-sm text-theme-secondary-500">
									<Amount ticker={exchangeTicker} value={displayValueConverted} />
								</div>
							)}
						</>
					)}
				</div>
			</ButtonGroupOption>
		))}
	</ButtonGroup>
);
