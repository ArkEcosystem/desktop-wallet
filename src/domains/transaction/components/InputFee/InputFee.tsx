import { InputRange } from "app/components/Input";
import { SelectionBar, SelectionBarOption } from "app/components/SelectionBar";
import React from "react";
import { useTranslation } from "react-i18next";

type InputFeeProps = {
	defaultValue: number;
	min: number;
	max: number;
	step: number;
	selectionBarState: any;
};

export const InputFee = ({ defaultValue, min, max, step, selectionBarState }: InputFeeProps) => {
	const { t } = useTranslation();

	return (
		<div className="flex space-x-2">
			<div className="flex-1">
				<InputRange name="fee" defaultValue={defaultValue} min={min} max={max} step={step} />
			</div>
			<div>
				<SelectionBar>
					<SelectionBarOption value={1} {...selectionBarState}>
						{t("TRANSACTION.FEES.LAST")}
					</SelectionBarOption>
					<SelectionBarOption value={2} {...selectionBarState}>
						{t("TRANSACTION.FEES.MIN")}
					</SelectionBarOption>
					<SelectionBarOption value={3} {...selectionBarState}>
						{t("TRANSACTION.FEES.AVERAGE")}
					</SelectionBarOption>
				</SelectionBar>
			</div>
		</div>
	);
};
