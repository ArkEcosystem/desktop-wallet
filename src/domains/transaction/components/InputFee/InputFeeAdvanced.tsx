import { InputRange } from "app/components/Input";
import React from "react";

interface Props {
	onChange: (value: string) => void;
	value: string;
	min: string;
	max: string;
	step: number;
	disabled: boolean;
}

export const InputFeeAdvanced = ({ onChange, value, min, max, step, disabled }: Props) => (
	<InputRange disabled={disabled} value={value} min={+min} max={+max} step={step} onChange={onChange} />
);
