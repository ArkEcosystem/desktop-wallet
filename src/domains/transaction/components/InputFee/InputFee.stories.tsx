import React, { useState } from "react";

import { InputFee } from "./InputFee";

export default { title: "App / Components / InputFee" };

export const Default = () => {
	const [value, setValue] = useState((5 * 1e8).toFixed(0));
	const feeOptions = {
		last: (2 * 1e8).toFixed(0),
		min: (1 * 1e8).toFixed(0),
		max: (10 * 1e8).toFixed(0),
		average: (1.354 * 1e8).toFixed(0),
	};

	return (
		<div className="max-w-lg">
			<InputFee {...feeOptions} defaultValue={value} step={0.01} onChange={(toshi: any) => setValue(toshi)} />
		</div>
	);
};
