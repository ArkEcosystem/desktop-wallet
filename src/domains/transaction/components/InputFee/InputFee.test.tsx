import { InputFee } from "domains/transaction/components/InputFee";
import React from "react";
import { act, fireEvent, render, waitFor } from "utils/testing-library";

import { translations as transactionTranslations } from "../../i18n";

let value: { display: string; value: string };
let min: string;
let max: string;
let avg: string;
let step: number;

let defaultProps = {};

describe("InputFee", () => {
	beforeEach(() => {
		defaultProps = {
			value: (2 * 1e8).toFixed(0),
			min: (1 * 1e8).toFixed(0),
			max: (10 * 1e8).toFixed(0),
			avg: (1.355 * 1e8).toFixed(0),
			step: 0.001,
			showFeeOptions: true,
		};
	});

	it("should render", () => {
		const { asFragment, getByTestId } = render(<InputFee {...defaultProps} defaultValue={defaultProps.value} />);

		expect(getByTestId("InputCurrency")).toHaveValue("2");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should update fee if value property changes", async () => {
		const { asFragment, getByTestId, rerender } = render(
			<InputFee {...defaultProps} defaultValue={defaultProps.value} />,
		);

		expect(getByTestId("InputCurrency")).toHaveValue("2");

		rerender(<InputFee {...defaultProps} defaultValue={defaultProps.value} value={"100000000"} />);

		await waitFor(() => {
			expect(getByTestId("InputCurrency")).toHaveValue("1");
		});

		expect(asFragment()).toMatchSnapshot();
	});

	it("should update fee when clicking on min", async () => {
		const onChange = jest.fn();

		const { asFragment, getByText, getByTestId } = render(
			<InputFee {...defaultProps} defaultValue={defaultProps.value} onChange={onChange} />,
		);
		expect(getByTestId("InputCurrency")).toHaveValue("2");

		act(() => {
			fireEvent.click(getByText(transactionTranslations.FEES.SLOW));
		});

		await waitFor(() => {
			expect(getByTestId("InputCurrency")).toHaveValue("1");
			expect(onChange).toHaveBeenCalledWith({ display: "1", value: "100000000" });
		});

		expect(asFragment()).toMatchSnapshot();
	});

	it("should update fee when clicking on avg", async () => {
		const onChange = jest.fn();

		const { asFragment, getByText, getByTestId } = render(
			<InputFee {...defaultProps} defaultValue={defaultProps.value} onChange={onChange} />,
		);

		act(() => {
			fireEvent.click(getByText(transactionTranslations.FEES.AVERAGE));
		});

		await waitFor(() => {
			expect(getByTestId("InputCurrency")).toHaveValue("1.355");
			expect(onChange).toHaveBeenCalledWith({ display: "1.355", value: "135500000" });
		});

		expect(asFragment()).toMatchSnapshot();
	});

	it("should update fee when clicking on max", async () => {
		const onChange = jest.fn();

		const { asFragment, getByText, getByTestId } = render(
			<InputFee {...defaultProps} defaultValue={defaultProps.value} onChange={onChange} />,
		);

		act(() => {
			fireEvent.click(getByText(transactionTranslations.FEES.FAST));
		});

		await waitFor(() => {
			expect(getByTestId("InputCurrency")).toHaveValue("10");
			expect(onChange).toHaveBeenCalledWith({ display: "10", value: "1000000000" });
		});

		expect(asFragment()).toMatchSnapshot();
	});
});
