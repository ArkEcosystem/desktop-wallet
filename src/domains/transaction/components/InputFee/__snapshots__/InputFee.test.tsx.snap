// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFee should render 1`] = `
<DocumentFragment>
  <div
    class="flex space-x-2"
    data-testid="InputFee"
  >
    <div
      class="flex-1"
    >
      <div
        class="InputGroup__InputGroupStyle-sc-1obbdqe-3 iqafwD"
      >
        <input
          class="Input__InputStyled-wu99qs-0 eVdxgh overflow-hidden w-full bg-theme-background appearance-none rounded border border-theme-secondary-400 dark:border-theme-secondary-700 text-theme-text transition-colors duration-200 px-4 py-3 no-ligatures"
          data-testid="InputCurrency"
          type="text"
          value="2"
        />
        <div
          class="absolute bottom-0 px-1 w-full"
        >
          <div
            class="flex flex-wrap justify-center"
            data-testid="Range"
          >
            <div
              class="Range__Track-sc-10itqc2-0 sgjTD"
              data-testid="Range__track"
              style="transform: scale(1); cursor: pointer;"
            >
              <div
                class="Range__TrackFilled-sc-10itqc2-1 egFfKP"
                data-testid="Range__track__filled"
              >
                <div
                  aria-valuemax="10"
                  aria-valuemin="1"
                  aria-valuenow="2"
                  class="Range__Thumb-sc-10itqc2-2 hXgqqU"
                  data-testid="Range__thumb"
                  draggable="false"
                  role="slider"
                  style="position: absolute; z-index: 0; cursor: grab; user-select: none; transform: translate(0px, 0px);"
                  tabindex="0"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="inline-flex overflow-hidden items-center space-x-2"
      data-testid="ButtonGroup"
      role="radiogroup"
    >
      <button
        aria-checked="false"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Slow
      </button>
      <button
        aria-checked="false"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Average
      </button>
      <button
        aria-checked="false"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Fast
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InputFee should update fee if value property changes 1`] = `
<DocumentFragment>
  <div
    class="flex space-x-2"
    data-testid="InputFee"
  >
    <div
      class="flex-1"
    >
      <div
        class="InputGroup__InputGroupStyle-sc-1obbdqe-3 iqafwD"
      >
        <input
          class="Input__InputStyled-wu99qs-0 eVdxgh overflow-hidden w-full bg-theme-background appearance-none rounded border border-theme-secondary-400 dark:border-theme-secondary-700 text-theme-text transition-colors duration-200 px-4 py-3 no-ligatures"
          data-testid="InputCurrency"
          type="text"
          value="1"
        />
        <div
          class="absolute bottom-0 px-1 w-full"
        >
          <div
            class="flex flex-wrap justify-center"
            data-testid="Range"
          >
            <div
              class="Range__Track-sc-10itqc2-0 sgjTD"
              data-testid="Range__track"
              style="transform: scale(1); cursor: pointer;"
            >
              <div
                class="Range__TrackFilled-sc-10itqc2-1 egFfKP"
                data-testid="Range__track__filled"
              >
                <div
                  aria-valuemax="10"
                  aria-valuemin="1"
                  aria-valuenow="1"
                  class="Range__Thumb-sc-10itqc2-2 hXgqqU"
                  data-testid="Range__thumb"
                  draggable="false"
                  role="slider"
                  style="position: absolute; z-index: 0; cursor: grab; user-select: none; transform: translate(0px, 0px);"
                  tabindex="0"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="inline-flex overflow-hidden items-center space-x-2"
      data-testid="ButtonGroup"
      role="radiogroup"
    >
      <button
        aria-checked="true"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Slow
      </button>
      <button
        aria-checked="false"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Average
      </button>
      <button
        aria-checked="false"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Fast
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InputFee should update fee when clicking on avg 1`] = `
<DocumentFragment>
  <div
    class="flex space-x-2"
    data-testid="InputFee"
  >
    <div
      class="flex-1"
    >
      <div
        class="InputGroup__InputGroupStyle-sc-1obbdqe-3 iqafwD"
      >
        <input
          class="Input__InputStyled-wu99qs-0 eVdxgh overflow-hidden w-full bg-theme-background appearance-none rounded border border-theme-secondary-400 dark:border-theme-secondary-700 text-theme-text transition-colors duration-200 px-4 py-3 no-ligatures"
          data-testid="InputCurrency"
          type="text"
          value="1.355"
        />
        <div
          class="absolute bottom-0 px-1 w-full"
        >
          <div
            class="flex flex-wrap justify-center"
            data-testid="Range"
          >
            <div
              class="Range__Track-sc-10itqc2-0 sgjTD"
              data-testid="Range__track"
              style="transform: scale(1); cursor: pointer;"
            >
              <div
                class="Range__TrackFilled-sc-10itqc2-1 egFfKP"
                data-testid="Range__track__filled"
              >
                <div
                  aria-valuemax="10"
                  aria-valuemin="1"
                  aria-valuenow="1.355"
                  class="Range__Thumb-sc-10itqc2-2 hXgqqU"
                  data-testid="Range__thumb"
                  draggable="false"
                  role="slider"
                  style="position: absolute; z-index: 0; cursor: grab; user-select: none; transform: translate(0px, 0px);"
                  tabindex="0"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="inline-flex overflow-hidden items-center space-x-2"
      data-testid="ButtonGroup"
      role="radiogroup"
    >
      <button
        aria-checked="false"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Slow
      </button>
      <button
        aria-checked="true"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Average
      </button>
      <button
        aria-checked="false"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Fast
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InputFee should update fee when clicking on max 1`] = `
<DocumentFragment>
  <div
    class="flex space-x-2"
    data-testid="InputFee"
  >
    <div
      class="flex-1"
    >
      <div
        class="InputGroup__InputGroupStyle-sc-1obbdqe-3 iqafwD"
      >
        <input
          class="Input__InputStyled-wu99qs-0 eVdxgh overflow-hidden w-full bg-theme-background appearance-none rounded border border-theme-secondary-400 dark:border-theme-secondary-700 text-theme-text transition-colors duration-200 px-4 py-3 no-ligatures"
          data-testid="InputCurrency"
          type="text"
          value="10"
        />
        <div
          class="absolute bottom-0 px-1 w-full"
        >
          <div
            class="flex flex-wrap justify-center"
            data-testid="Range"
          >
            <div
              class="Range__Track-sc-10itqc2-0 sgjTD"
              data-testid="Range__track"
              style="transform: scale(1); cursor: pointer;"
            >
              <div
                class="Range__TrackFilled-sc-10itqc2-1 egFfKP"
                data-testid="Range__track__filled"
              >
                <div
                  aria-valuemax="10"
                  aria-valuemin="1"
                  aria-valuenow="10"
                  class="Range__Thumb-sc-10itqc2-2 hXgqqU"
                  data-testid="Range__thumb"
                  draggable="false"
                  role="slider"
                  style="position: absolute; z-index: 0; cursor: grab; user-select: none; transform: translate(0px, 0px);"
                  tabindex="0"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="inline-flex overflow-hidden items-center space-x-2"
      data-testid="ButtonGroup"
      role="radiogroup"
    >
      <button
        aria-checked="false"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Slow
      </button>
      <button
        aria-checked="false"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Average
      </button>
      <button
        aria-checked="true"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Fast
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InputFee should update fee when clicking on min 1`] = `
<DocumentFragment>
  <div
    class="flex space-x-2"
    data-testid="InputFee"
  >
    <div
      class="flex-1"
    >
      <div
        class="InputGroup__InputGroupStyle-sc-1obbdqe-3 iqafwD"
      >
        <input
          class="Input__InputStyled-wu99qs-0 eVdxgh overflow-hidden w-full bg-theme-background appearance-none rounded border border-theme-secondary-400 dark:border-theme-secondary-700 text-theme-text transition-colors duration-200 px-4 py-3 no-ligatures"
          data-testid="InputCurrency"
          type="text"
          value="1"
        />
        <div
          class="absolute bottom-0 px-1 w-full"
        >
          <div
            class="flex flex-wrap justify-center"
            data-testid="Range"
          >
            <div
              class="Range__Track-sc-10itqc2-0 sgjTD"
              data-testid="Range__track"
              style="transform: scale(1); cursor: pointer;"
            >
              <div
                class="Range__TrackFilled-sc-10itqc2-1 egFfKP"
                data-testid="Range__track__filled"
              >
                <div
                  aria-valuemax="10"
                  aria-valuemin="1"
                  aria-valuenow="1"
                  class="Range__Thumb-sc-10itqc2-2 hXgqqU"
                  data-testid="Range__thumb"
                  draggable="false"
                  role="slider"
                  style="position: absolute; z-index: 0; cursor: grab; user-select: none; transform: translate(0px, 0px);"
                  tabindex="0"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="inline-flex overflow-hidden items-center space-x-2"
      data-testid="ButtonGroup"
      role="radiogroup"
    >
      <button
        aria-checked="true"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Slow
      </button>
      <button
        aria-checked="false"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Average
      </button>
      <button
        aria-checked="false"
        class="ButtonGroup__ButtonGroupOptionStyled-sc-1u00raz-0 cQPjbG"
        data-testid="ButtonGroupOption"
        role="radio"
        type="button"
      >
        Fast
      </button>
    </div>
  </div>
</DocumentFragment>
`;
